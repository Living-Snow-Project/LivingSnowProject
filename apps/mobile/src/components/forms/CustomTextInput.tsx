import React, { forwardRef /* useRef */ } from "react";
import { FormControl, Input, WarningOutlineIcon } from "native-base";
/* import {
  NativeSyntheticEvent,
  TextInputContentSizeChangeEventData,
} from "react-native";
import RCTDeviceEventEmitter from "react-native/Libraries/EventEmitter/RCTDeviceEventEmitter"; */

type CustomTextInputProps = {
  label: string;
  placeholder: string;
  value?: string | undefined;
  validation?: string;
  maxLength?: number;
  isDisabled?: boolean;
  isInvalid?: boolean;
  isRequired?: boolean;
  onChangeText?: (text: string) => void;
  onSubmitEditing?: () => void;
};

export const CustomTextInput = forwardRef<typeof Input, CustomTextInputProps>(
  (
    {
      label,
      placeholder,
      value,
      validation,
      maxLength = 255,
      isDisabled = false,
      isInvalid = false,
      isRequired = false,
      onChangeText = () => {},
      onSubmitEditing = () => {},
    }: CustomTextInputProps,
    ref
  ) => (
    /* const textInputHeight = useRef(0);

    // onContentSizeChange is called frequently for multiline TextInput, we only want to emit 'keyboardDidShow' event when height actually changes
    const handleMultilineTextInputOnContentSizeChange = (
      height: number,
      event: NativeSyntheticEvent<TextInputContentSizeChangeEventData>
    ) => {
      if (height !== 0 && height !== event.nativeEvent.contentSize.height) {
        // 'keyboardDidShow' expects the height of the keyboard (which we could capture in a new event listener in this component)
        // since we only have 1 'keyboardDidShow' listener we changed its logic to respond to this input
        // this is potentially bad if we set up additional 'keyboardDidShow' listeners in the app
        RCTDeviceEventEmitter.emit("keyboardDidShow", {});
      }

      return event.nativeEvent.contentSize.height;
    }; */

    <FormControl isRequired={isRequired} isInvalid={isInvalid}>
      <FormControl.Label>{label}</FormControl.Label>
      <Input
        size="lg"
        multiline
        blurOnSubmit
        // @ts-ignore ref typing isn't quite right in native-base yet
        ref={ref}
        value={value}
        placeholder={placeholder}
        isDisabled={isDisabled}
        onChangeText={onChangeText}
        onSubmitEditing={onSubmitEditing}
        /* onContentSizeChange={(event) => {
          textInputHeight.current =
            handleMultilineTextInputOnContentSizeChange(
              textInputHeight.current,
              event
            );
        }} */
        maxLength={maxLength}
        returnKeyType="done"
        variant="outline"
      />
      <FormControl.ErrorMessage leftIcon={<WarningOutlineIcon size="sm" />}>
        {validation}
      </FormControl.ErrorMessage>
    </FormControl>
  )
);

// this is needed because of forwardRef
CustomTextInput.displayName = `CustomTextInput`;
