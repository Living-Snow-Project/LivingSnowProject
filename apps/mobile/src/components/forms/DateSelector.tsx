import React, { useState } from "react";
import { View } from "react-native";
import { FormControl, Input, Pressable, useColorModeValue } from "native-base";
import { Calendar } from "react-native-calendars";
import { Labels } from "../../constants";

type DateSelectorProps = {
  date: string;
  maxDate: string;
  setDate: (date: string) => void;
};

const dark = "#44403c"; // NativeBase light.700
const light = "#fafaf9"; // NativeBase light.50

export function DateSelector({ date, maxDate, setDate }: DateSelectorProps) {
  const [calendarVisible, setCalendarVisible] = useState(false);
  const bgColor = useColorModeValue(light, dark);
  const dayColor = useColorModeValue(dark, light);
  const selectedDayColor = bgColor;
  const monthColor = dayColor;
  const disabledColor = useColorModeValue(`${dark}55`, `${light}55`);

  const renderCalendar = () => {
    if (calendarVisible) {
      return (
        <Calendar
          testID="calendar"
          current={date}
          onDayPress={(newDate) => {
            setCalendarVisible(false);
            setDate(newDate.dateString);
          }}
          maxDate={maxDate}
          markedDates={{ [date]: { selected: true } }}
          theme={{
            calendarBackground: bgColor,
            dayTextColor: dayColor,
            selectedDayTextColor: selectedDayColor,
            monthTextColor: monthColor,
            textDisabledColor: disabledColor,
          }}
        />
      );
    }

    return (
      <View pointerEvents="none">
        <Input value={date} size="lg" />
      </View>
    );
  };

  return (
    <FormControl isRequired>
      <FormControl.Label>{Labels.Date}</FormControl.Label>
      <Pressable
        testID="calendar-pressable"
        onPress={() => setCalendarVisible(true)}
      >
        {renderCalendar()}
      </Pressable>
    </FormControl>
  );
}
