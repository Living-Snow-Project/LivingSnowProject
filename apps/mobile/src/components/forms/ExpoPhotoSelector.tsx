import React from "react";
import { FormControl, Pressable } from "native-base";
import * as ImagePicker from "expo-image-picker";
import * as MediaLibrary from "expo-media-library";
import { manipulateAsync, ActionResize } from "expo-image-manipulator";
import { SelectedPhoto } from "../../../types";
import { ThemedBox } from "../layout";
import { AddPhotosIcon, PhotosLayout } from "../media";
import { PhotoManager } from "../../lib/PhotoManager";
import { Labels, TestIds } from "../../constants";
import { useToast } from "../../hooks";
import { ToastAlert } from "../feedback";

type ExpoPhotoSelectorProps = {
  recordId: string;
  photos: SelectedPhoto[];
  setSelectedPhotos: (selectedPhotos: SelectedPhoto[]) => void;
  setStatus: (status: "Idle" | "Loading") => void;
};

export function ExpoPhotoSelector({
  recordId,
  photos,
  setSelectedPhotos,
  setStatus,
}: ExpoPhotoSelectorProps) {
  const toast = useToast();

  const handleOnPress = async () => {
    setStatus("Loading");

    try {
      const permission = await MediaLibrary.getPermissionsAsync();

      if (!permission.granted) {
        const granted = await MediaLibrary.requestPermissionsAsync();

        if (!granted.granted) {
          return;
        }
      }

      const result = await ImagePicker.launchImageLibraryAsync({
        mediaTypes: ImagePicker.MediaTypeOptions.Images,
        allowsMultipleSelection: true,
        selectionLimit: 4,
        quality: 0.7,
      });

      // TODO: should "cancel" unselect all? there is no way to go back to 0 selected after first selection
      if (!result.canceled) {
        const assets: MediaLibrary.Asset[] = [];

        await result.assets.reduce(async (promise, current) => {
          await promise;

          if (current.assetId) {
            // returned ImagePickerAsset is missing various fields in Asset that are "needed" (for TypeScript)
            const asset = await MediaLibrary.getAssetInfoAsync(current.assetId);

            // resize to 1024 along major axis
            const dims: ActionResize = { resize: {} };

            if (asset.height > asset.width) {
              dims.resize.height = 1024;
            } else {
              dims.resize.width = 1024;
            }

            const resizedImage = await manipulateAsync(current.uri, [dims]);

            assets.push({
              ...asset,
              height: resizedImage.height,
              uri: resizedImage.uri,
              width: resizedImage.width,
            });
          }

          return Promise.resolve();
        }, Promise.resolve());

        PhotoManager.addSelected(recordId, assets);
        setSelectedPhotos(assets);
      }
    } catch (error) {
      toast.show(
        <ToastAlert
          status="info"
          title="Upload photo error"
          message="We ran into an error preparing photos for upload."
        />,
      );
    } finally {
      setStatus("Idle");
    }
  };

  const renderPhotos = () => {
    if (photos.length) {
      return (
        <ThemedBox mx={-2}>
          <PhotosLayout photos={photos} />
        </ThemedBox>
      );
    }

    return (
      <ThemedBox height="52" width="52">
        <AddPhotosIcon />
      </ThemedBox>
    );
  };

  return (
    <FormControl>
      <FormControl.Label>{Labels.RecordScreen.Photos}</FormControl.Label>
      <Pressable testID={TestIds.Selectors.Photos} onPress={handleOnPress}>
        {renderPhotos()}
      </Pressable>
    </FormControl>
  );
}
