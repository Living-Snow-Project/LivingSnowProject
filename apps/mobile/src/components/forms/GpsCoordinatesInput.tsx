import React, { useEffect, useRef, useState } from "react";
import {
  Alert,
  Pressable,
  ReturnKeyTypeOptions,
  Text,
  TextInput,
  View,
} from "react-native";
import {
  Accuracy,
  requestForegroundPermissionsAsync,
  watchPositionAsync,
} from "expo-location";
import Logger from "@livingsnow/logger";
import { formInputStyles } from "../../styles/FormInput";
import { getAppSettings } from "../../../AppSettings";
import { TestIds } from "../../constants/TestIds";
import { Placeholders } from "../../constants/Strings";

type GpsCoordinates = {
  latitude: number;
  longitude: number;
};

type GpsCoordinatesInputProps = {
  setGpsCoordinates: (latitude: number, longitude: number) => void;
  onSubmitEditing: () => void;
  coordinates?: GpsCoordinates;
};

export function GpsCoordinatesInput({
  setGpsCoordinates,
  onSubmitEditing,
  coordinates = { latitude: 0, longitude: 0 },
}: GpsCoordinatesInputProps) {
  const watchPosition = useRef<null | { remove(): void }>(null);
  const gpsCoordinatesRef = useRef<TextInput>(null);
  const [placeholderText, setPlaceholderText] = useState(
    Placeholders.GPS.AcquiringLocation
  );
  const useGps = coordinates.latitude === 0 && coordinates.longitude === 0;
  const [gpsCoordinateString, setGpsCoordinateString] = useState<string>(
    useGps ? `` : `${coordinates.latitude}, ${coordinates.longitude}`
  );
  const [manualGpsCoordinates, setManualGpsCoordinates] = useState(!useGps);
  const showGpsWarning = useGps ? getAppSettings().showGpsWarning : false;
  const clipCoordinate = (coordinate: number) =>
    JSON.stringify(coordinate.toFixed(6)).replace('"', "").replace('"', "");

  // accepts inputs from both location subscription and typed user input
  const updateGpsCoordinateString = (value: string) => {
    setGpsCoordinateString(value);

    // some users are adding parenthesis when manually entering coordinates
    // we don't want them in the data set so quietly remove
    const coords = value.replace(`(`, ``).replace(`)`, ``).split(",");

    let latitude;
    let longitude;

    if (coords[0]) {
      latitude = Number(coords[0].trim());
    }

    if (coords[1]) {
      longitude = Number(coords[1].trim());
    }

    setGpsCoordinates(latitude, longitude);
  };

  const stopGps = () => {
    if (watchPosition.current) {
      watchPosition.current?.remove();
      watchPosition.current = null;
    }
  };

  // location subscription
  useEffect(() => {
    if (!useGps) return () => {};

    (async () => {
      const { status } = await requestForegroundPermissionsAsync();
      if (status !== "granted") {
        setManualGpsCoordinates(true);
        setPlaceholderText(Placeholders.GPS.NoPermissions);
        Logger.Warn(`Permission to access foreground location was ${status}.`);
        return;
      }

      try {
        watchPosition.current = await watchPositionAsync(
          {
            accuracy: Accuracy.High,
            timeInterval: 5000,
          },
          ({ coords }) =>
            updateGpsCoordinateString(
              `${clipCoordinate(coords.latitude)}, ${clipCoordinate(
                coords.longitude
              )}`
            )
        );
      } catch (error) {
        setManualGpsCoordinates(true);
        setPlaceholderText(Placeholders.GPS.NoLocation);
        Logger.Error(`Location.watchPositionAsync() failed: ${error}`);
      }
    })();

    return stopGps;
  }, []);

  useEffect(() => {
    if (useGps) {
      gpsCoordinatesRef?.current?.focus();
    }
  }, [manualGpsCoordinates]);

  const enableManualGpsCoordinates = () => {
    stopGps();
    updateGpsCoordinateString("");
    setManualGpsCoordinates(true);
    setPlaceholderText(Placeholders.GPS.EnterCoordinates);
  };

  const confirmManualGpsCoordinates = () => {
    if (showGpsWarning) {
      Alert.alert(
        "Enter GPS coordinates manually?",
        "This message can be disabled in Settings.",
        [
          {
            text: "Yes",
            onPress: () => enableManualGpsCoordinates(),
          },
          {
            text: "No",
            style: "cancel",
          },
        ]
      );
    } else {
      enableManualGpsCoordinates();
    }
  };

  const gpsFieldProps = {
    style: formInputStyles.optionInputText,
    editable: manualGpsCoordinates,
    maxLength: 30,
    placeholder: placeholderText,
    defaultValue: gpsCoordinateString,
    returnKeyType: "done" as ReturnKeyTypeOptions,
  };

  return (
    <>
      <Text style={formInputStyles.optionStaticText}>
        GPS Coordinates (latitude, longitude)
      </Text>
      {!manualGpsCoordinates && (
        /* user confirms they want to enter GPS coordinates manually */
        <Pressable
          onPress={() => confirmManualGpsCoordinates()}
          testID={TestIds.GPS.gpsManualPressableTestId}
        >
          <View pointerEvents="none">
            <TextInput {...gpsFieldProps} />
          </View>
        </Pressable>
      )}
      {manualGpsCoordinates && (
        <TextInput
          {...gpsFieldProps}
          testID={TestIds.GPS.gpsManualInputTestId}
          onChangeText={(value) => updateGpsCoordinateString(value)}
          onSubmitEditing={() => onSubmitEditing()}
          ref={gpsCoordinatesRef}
        />
      )}
    </>
  );
}
