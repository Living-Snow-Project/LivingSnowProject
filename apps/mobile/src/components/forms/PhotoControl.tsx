import React, { useCallback } from "react";
import { Image, LogBox, Pressable, StyleSheet, Text, View } from "react-native";
import { SelectedPhoto } from "@livingsnow/record";
import { RootStackNavigationProp } from "../../navigation/Routes";
import { formInputStyles } from "../../styles/FormInput";
import TestIds from "../../constants/TestIds";
import { PictureIcon } from "../Icons";

// because we pass a callback in params, more info from the following links
// follow links for best practices, look at Context
// https://reactnavigation.org/docs/troubleshooting/#i-get-the-warning-non-serializable-values-were-found-in-the-navigation-state
// https://reactnavigation.org/docs/params/
// https://reactnavigation.org/docs/hello-react-navigation/#passing-additional-props
LogBox.ignoreLogs([
  "Non-serializable values were found in the navigation state",
]);

const styles = StyleSheet.create({
  // these styles could be dynamically sized\arranged based on number of photos and combinaions of portrait\landscape
  container: {
    flex: 1,
    margin: 1,
    height: 100,
  },
  photo: {
    width: "100%",
    height: "100%",
    borderColor: "black",
    borderWidth: 1,
    borderRadius: 2,
  },
});

type PhotoControlProps = {
  navigation: RootStackNavigationProp;
  photos: SelectedPhoto[];
  onUpdatePhotos: (photos: SelectedPhoto[]) => void;
};

export default function PhotoControl({
  navigation,
  photos,
  onUpdatePhotos,
}: PhotoControlProps) {
  const renderPhotos = useCallback(() => {
    const colsPerRow = 2;

    // layout algorithm for even number of photos
    if (photos.length % colsPerRow === 0) {
      const result: JSX.Element[] = [];
      const rows = photos.length / colsPerRow;

      for (let row = 0; row < rows; row++) {
        const inner: JSX.Element[] = [];
        for (let col = 0; col < colsPerRow; col++) {
          const index = row * rows + col;
          inner.push(
            <View style={styles.container} key={index}>
              <Image style={styles.photo} source={{ uri: photos[index].uri }} />
            </View>
          );
        }

        result.push(
          <View style={{ flex: rows, flexDirection: "row" }} key={row}>
            {inner}
          </View>
        );
      }

      // short circuit function
      return result;
    }

    // layout algorithm for odd number of photos
    return (
      <View style={{ flex: photos.length, flexDirection: "row" }}>
        {photos.map((x, index) => (
          /* eslint-disable react/no-array-index-key */
          <View style={styles.container} key={index}>
            <Image style={styles.photo} source={{ uri: x.uri }} />
          </View>
        ))}
      </View>
    );
  }, [photos]);

  return (
    <>
      <Text style={formInputStyles.optionStaticText}>
        Select Photos (limit 4)
      </Text>
      <Pressable
        testID={TestIds.Photos.photoSelectorTestId}
        onPress={() => {
          let existingSelection: string[] | undefined;

          if (photos.length > 0) {
            existingSelection = [];
            for (let i = 0; i < photos.length; i++) {
              existingSelection.push(photos[i].id);
            }
          }

          navigation.navigate("ImageSelection", {
            existingSelection,
            onUpdatePhotos,
          });
        }}
      >
        {photos.length === 0 && <PictureIcon />}
        {photos.length > 0 && renderPhotos()}
      </Pressable>
    </>
  );
}
