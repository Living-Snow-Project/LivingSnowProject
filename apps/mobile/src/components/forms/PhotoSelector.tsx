import React from "react";
import { FormControl, Pressable } from "native-base";
import { SelectedPhoto } from "@livingsnow/record";
import { RootStackNavigationProp } from "../../navigation/Routes";
import { ThemedBox } from "../ThemedBox";
import { AddPhotosIcon } from "../Icons";
import { Labels, TestIds } from "../../constants";
import { PhotosLayout } from "../PhotosLayout";

type PhotoSelectorProps = {
  navigation: RootStackNavigationProp;
  photos: SelectedPhoto[];
};

export function PhotoSelector({ navigation, photos }: PhotoSelectorProps) {
  const handleOnPress = () => {
    let existingSelection: string[] | undefined;

    if (photos.length > 0) {
      existingSelection = [];
      for (let i = 0; i < photos.length; i++) {
        existingSelection.push(photos[i].id);
      }
    }

    navigation.navigate("ImageSelection", { existingSelection });
  };

  const renderPhotos = () => {
    if (photos.length) {
      return (
        <ThemedBox mx={-2}>
          <PhotosLayout photos={photos} />
        </ThemedBox>
      );
    }

    return (
      <ThemedBox height="52" width="52">
        <AddPhotosIcon />
      </ThemedBox>
    );
  };

  return (
    <FormControl>
      <FormControl.Label>{Labels.RecordScreen.Photos}</FormControl.Label>
      <Pressable testID={TestIds.Selectors.Photos} onPress={handleOnPress}>
        {renderPhotos()}
      </Pressable>
    </FormControl>
  );
}
