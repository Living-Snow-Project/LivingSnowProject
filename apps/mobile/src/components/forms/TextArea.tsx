import React, { forwardRef } from "react";
import { FormControl, TextArea as NBTextArea } from "native-base";

type TextAreaProps = {
  label: string;
  placeholder: string;
  value?: string | undefined;
  maxLength?: number;
  blurOnSubmit?: boolean;
  onChangeText?: (text: string) => void;
  onSubmitEditing?: () => void;
};

export const TextArea = forwardRef<typeof NBTextArea, TextAreaProps>(
  (
    {
      label,
      placeholder,
      value,
      maxLength = 255,
      blurOnSubmit = false,
      onChangeText = () => {},
      onSubmitEditing,
    }: TextAreaProps,
    ref
  ) => (
    <FormControl>
      <FormControl.Label>{label}</FormControl.Label>
      {/* @ts-ignore missing typing for autoCompleteType */}
      <NBTextArea
        size="lg"
        blurOnSubmit={blurOnSubmit}
        // @ts-ignore ref typing isn't quite right in native-base yet
        ref={ref}
        value={value}
        placeholder={placeholder}
        onChangeText={(text) => {
          if (text.length && text[text.length - 1] == "\n") {
            return;
          }

          onChangeText(text);
        }}
        onSubmitEditing={onSubmitEditing}
        maxLength={maxLength}
        returnKeyType="done"
        variant="outline"
      />
    </FormControl>
  )
);

// this is needed because of forwardRef
TextArea.displayName = `TextArea`;
