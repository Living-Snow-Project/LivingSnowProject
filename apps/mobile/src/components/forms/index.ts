export * from "./CustomTextInput";
export * from "./DateSelector";
export * from "./ExpoPhotoSelector";
export * from "./GpsCoordinatesInput";
export * from "./PhotoSelector";
export * from "./TextArea";
export * from "./TypeSelector";
export * from "./UserIdentityInput";
