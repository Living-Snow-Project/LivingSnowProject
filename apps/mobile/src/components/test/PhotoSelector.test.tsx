import React from "react";
import { render, waitFor } from "@testing-library/react-native";
import { NativeBaseProviderForTesting } from "../../../jesttest.setup";
import { SelectedPhoto } from "../../../types";
import { RootStackNavigationProp } from "../../navigation/Routes";
import { PhotoSelector } from "../forms/PhotoSelector";

const navigation: RootStackNavigationProp = {} as RootStackNavigationProp;
navigation.goBack = () => {};
navigation.navigate = () => {};

describe("PhotoSelector test suite", () => {
  test("even number of photos", () => {
    const photos: SelectedPhoto[] = [
      { id: "0", uri: "", height: 16, width: 24 } as SelectedPhoto,
      { id: "1", uri: "", height: 24, width: 16 } as SelectedPhoto,
    ];

    const { getByText, toJSON } = render(
      <NativeBaseProviderForTesting>
        <PhotoSelector navigation={navigation} photos={photos} />
      </NativeBaseProviderForTesting>
    );

    waitFor(() => getByText("Select Photos (limit 4)"));

    expect(toJSON()).toMatchSnapshot();
  });

  test("odd number of photos", () => {
    const photos: SelectedPhoto[] = [
      { id: "0", uri: "", height: 16, width: 24 } as SelectedPhoto,
    ];

    const { getByText, toJSON } = render(
      <NativeBaseProviderForTesting>
        <PhotoSelector navigation={navigation} photos={photos} />
      </NativeBaseProviderForTesting>
    );

    waitFor(() => getByText("Select Photos (limit 4)"));

    expect(toJSON()).toMatchSnapshot();
  });
});
