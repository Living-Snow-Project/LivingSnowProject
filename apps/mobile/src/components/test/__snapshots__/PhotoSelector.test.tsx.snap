// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoSelector test suite even number of photos 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={{}}
    style={
      {
        "width": "100%",
      }
    }
  >
    <View
      dataSet={{}}
      feedbackId="field-1-feedback"
      hasFeedbackText={false}
      hasHelpText={false}
      helpTextId="field-1-helptext"
      isDisabled={false}
      isInvalid={false}
      isReadOnly={false}
      isRequired={false}
      labelId="field-1-label"
      nativeID="field-1-label"
      setHasFeedbackText={[Function]}
      setHasHelpText={[Function]}
      style={
        {
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "marginBottom": 4,
          "marginTop": 4,
        }
      }
    >
      <Text
        dataSet={{}}
        style={
          {
            "backgroundColor": undefined,
            "color": "#737373",
            "fontFamily": undefined,
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "500",
            "letterSpacing": 0,
            "lineHeight": 21,
            "textDecorationLine": undefined,
          }
        }
      >
        Select Photos (limit 4)
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      dataSet={{}}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={{}}
      testID="photo-selector"
    >
      <View
        dataSet={{}}
        style={
          {
            "marginLeft": -8,
            "marginRight": -8,
          }
        }
      >
        <View
          dataSet={{}}
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            dataSet={{}}
            style={
              {
                "alignItems": "center",
                "height": 372,
                "justifyContent": "center",
                "width": 248,
              }
            }
          >
            <ActivityIndicator
              accessibilityLabel="loading"
              accessible={true}
              color="#0891b2"
              size="large"
              style={
                [
                  [
                    {},
                    {
                      "dataSet": {},
                    },
                    {},
                  ],
                  undefined,
                ]
              }
            />
            <Text
              dataSet={{}}
              style={
                {
                  "backgroundColor": undefined,
                  "color": "#171717",
                  "fontFamily": undefined,
                  "fontSize": 14,
                  "fontStyle": "normal",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                  "lineHeight": 21,
                  "textDecorationLine": undefined,
                }
              }
            >
              Loading
            </Text>
          </View>
          <View
            dataSet={{}}
            style={{}}
          />
          <View
            dataSet={{}}
            style={
              {
                "paddingRight": 6,
              }
            }
          />
          <View
            dataSet={{}}
            style={{}}
          />
          <View
            dataSet={{}}
            style={
              {
                "alignItems": "center",
                "height": 372,
                "justifyContent": "center",
                "width": 496,
              }
            }
          >
            <ActivityIndicator
              accessibilityLabel="loading"
              accessible={true}
              color="#0891b2"
              size="large"
              style={
                [
                  [
                    {},
                    {
                      "dataSet": {},
                    },
                    {},
                  ],
                  undefined,
                ]
              }
            />
            <Text
              dataSet={{}}
              style={
                {
                  "backgroundColor": undefined,
                  "color": "#171717",
                  "fontFamily": undefined,
                  "fontSize": 14,
                  "fontStyle": "normal",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                  "lineHeight": 21,
                  "textDecorationLine": undefined,
                }
              }
            >
              Loading
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`PhotoSelector test suite odd number of photos 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={{}}
    style={
      {
        "width": "100%",
      }
    }
  >
    <View
      dataSet={{}}
      feedbackId="field-2-feedback"
      hasFeedbackText={false}
      hasHelpText={false}
      helpTextId="field-2-helptext"
      isDisabled={false}
      isInvalid={false}
      isReadOnly={false}
      isRequired={false}
      labelId="field-2-label"
      nativeID="field-2-label"
      setHasFeedbackText={[Function]}
      setHasHelpText={[Function]}
      style={
        {
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "marginBottom": 4,
          "marginTop": 4,
        }
      }
    >
      <Text
        dataSet={{}}
        style={
          {
            "backgroundColor": undefined,
            "color": "#737373",
            "fontFamily": undefined,
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "500",
            "letterSpacing": 0,
            "lineHeight": 21,
            "textDecorationLine": undefined,
          }
        }
      >
        Select Photos (limit 4)
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      dataSet={{}}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={{}}
      testID="photo-selector"
    >
      <View
        dataSet={{}}
        style={
          {
            "marginLeft": -8,
            "marginRight": -8,
          }
        }
      >
        <View
          dataSet={{}}
          style={
            {
              "alignItems": "center",
              "height": 500,
              "justifyContent": "center",
              "width": 750,
            }
          }
        >
          <ActivityIndicator
            accessibilityLabel="loading"
            accessible={true}
            color="#0891b2"
            size="large"
            style={
              [
                [
                  {},
                  {
                    "dataSet": {},
                  },
                  {},
                ],
                undefined,
              ]
            }
          />
          <Text
            dataSet={{}}
            style={
              {
                "backgroundColor": undefined,
                "color": "#171717",
                "fontFamily": undefined,
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "400",
                "letterSpacing": 0,
                "lineHeight": 21,
                "textDecorationLine": undefined,
              }
            }
          >
            Loading
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
